gtsfm/utils/metrics.py

```diff
@@ -59,15 +64,145 @@ def count_correct_correspondences(
 
+    Returns:
+        is_inlier: (N, ) mask of inlier correspondences.
+        distance_squared: squared sampson distance between corresponding keypoints.
+    """
     i2Ei1 = EssentialMatrix(i2Ti1.rotation(), Unit3(i2Ti1.translation()))
     i2Fi1 = verification_utils.essential_to_fundamental_matrix(i2Ei1, intrinsics_i1, intrinsics_i2)
-
     distance_squared = verification_utils.compute_epipolar_distances_sq_sampson(
         keypoints_i1.coordinates, keypoints_i2.coordinates, i2Fi1
     )
-    return np.count_nonzero(distance_squared < epipolar_dist_threshold ** 2)
+    is_inlier = distance_squared < dist_threshold ** 2 if distance_squared is not None else None
+
+    return is_inlier, distance_squared
+
+
+def mesh_inlier_correspondences(
+    keypoints_i1: Keypoints,
+    keypoints_i2: Keypoints,
+    gt_camera_i1: PinholeCameraCal3Bundler,
+    gt_camera_i2: PinholeCameraCal3Bundler,
+    gt_scene_mesh: trimesh.Trimesh,
+    dist_threshold: float,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Compute inlier correspondences using the ground truth triangular surface mesh of the scene.
+
+    Args:
+        keypoints_i1: N keypoints in image i1.
+        keypoints_i2: N corresponding keypoints in image i2.
+        intrinsics_i1: intrinsics for i1.
+        intrinsics_i2: intrinsics for i2.
+        gt_wTi1: ground truth pose of the world frame relative to i1.
+        gt_wTi2: ground truth pose of the world frame relative to i2.
+        gt_scene_mesh: ground truth triangular surface mesh of the scene in the world frame.
+        dist_threshold: max acceptable distance for a correct correspondence.
+
+    Returns:
+        is_inlier: (N, ) mask of inlier correspondences.
+        reproj_err: maximum error between forward-projected ground truth landmark and corresponding keypoints
+
+    Raises:
+        ValueError if the number of keypoints do not match.
+    """
+    if len(keypoints_i1) != len(keypoints_i2):
+        raise ValueError("Keypoints must have same counts")
+
+    n_corrs = len(keypoints_i1)
+    is_inlier = np.zeros(n_corrs, dtype=bool)
+    src_i1 = np.repeat(np.reshape(gt_camera_i1.translation(), (-1, 3)), n_corrs, axis=0)
+    src_i2 = np.repeat(np.reshape(gt_camera_i2.translation(), (-1, 3)), n_corrs, axis=0)
+
+    # Compute ketpoint rays.
+    drc_i1 = np.empty((n_corrs, 3), dtype=float)
+    drc_i2 = np.empty((n_corrs, 3), dtype=float)
+    for corr_idx in range(n_corrs):
+        drc_i1[corr_idx, :] = (
+            gt_camera_i1.BackprojectFromCamera(keypoints_i1.coordinates[corr_idx], depth=1.0)
+            - gt_camera_i1.translation()
+        )
+        drc_i2[corr_idx, :] = (
+            gt_camera_i2.BackprojectFromCamera(keypoints_i2.coordinates[corr_idx], depth=1.0)
+            - gt_camera_i2.translation()
+        )
+
+    # Perform ray tracing.
+    src = np.vstack((src_i1, src_i2))
+    drc = np.vstack((drc_i1, drc_i2))
+    logger.info("Computing ray intersections...")
```

Comment:
Nit: this logging could get expensive for 50K+ image pairs, like skydio-501 w/ 150 frame lookahead